### PR TITLE
feat/update-execution-labels-at-runtime

### LINKED ISSUE #2632: Add a dedicated task allowing to set custom Execution Labels from a flow
### Feature description

Possible syntax:

```yaml
id: labels_test
namespace: dev

inputs:
  - name: myinput
    type: STRING

tasks:
  - id: labels
    type: io.kestra.core.tasks.executions.Labels
    labels:
      mykey: "{{ inputs.myinput }}"
```

This will also address this request in a straightforward and explicit manner https://github.com/kestra-io/kestra/issues/1760 as it's otherwise difficult to sometimes add labels from inputs (nested JSON etc)

Additionally, this task allows setting execution labels based on specific task outputs, flow progress, etc. - many use cases enabled by adding this.

### PR DESCRIPTION
**Pull request by brian-mulier-p**
closes #2632

### CODE REVIEW

File: core/src/main/java/io/kestra/core/models/tasks/ExecutionUpdatingTask.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package io.kestra.core.models.tasks;
+
+import io.kestra.core.models.executions.Execution;
+import io.kestra.core.runners.RunContext;
+
+/**
+ * Interface for tasks that modify the execution at runtime.
+ */
+public interface ExecutionUpdatingTask {
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/models/tasks/ExecutionUpdatingTask.java:9]: All other task types are suffixed by _able_, maybe call it ExecutionUpdatableTask ?

File: core/src/main/java/io/kestra/core/tasks/executions/Labels.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+import static io.kestra.core.utils.Rethrow.throwBiConsumer;
+import static io.kestra.core.utils.Rethrow.throwFunction;
+
+@SuperBuilder
+@ToString
+@EqualsAndHashCode
+@Getter
+@NoArgsConstructor
+@Schema(
+    title = "Allow to add labels to current execution at runtime.",
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/tasks/executions/Labels.java:38]:
```suggestion
    title = "Allow to add labels to the current execution at runtime.",
```

File: core/src/main/java/io/kestra/core/tasks/executions/Labels.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    description = "Can be used to ease filtering executions by some runtime-fetched data."
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/tasks/executions/Labels.java:39]: I don't think this description brings much value

File: core/src/main/java/io/kestra/core/tasks/executions/Labels.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+)
+@Plugin(
+    examples = {
+        @Example(
+            title = "Add labels based on a webhook payload",
+            full = true,
+            code = {
+                "id: webhook-based-labels",
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/tasks/executions/Labels.java:47]: Please use text block for such long string.
Don't modify it but think for the next time.

File: core/src/main/java/io/kestra/core/tasks/executions/Labels.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                "    conditions:",
+                "      - type: io.kestra.core.models.conditions.types.VariableCondition",
+                "        expression: \"{{ trigger.body.customerId is defined and trigger.body.orderId is defined and trigger.body.orderType is defined }}\""
+            }
+        )
+    }
+)
+public class Labels extends Task implements ExecutionUpdatingTask {
+    @Schema(
+        title = "Labels to add to current execution.",
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/tasks/executions/Labels.java:74]:
```suggestion
        title = "Labels to add to the current execution.",
```

File: core/src/main/java/io/kestra/core/tasks/executions/Labels.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        description = "The value should result in a list of labels or a labelKey:labelValue map",
+        anyOf = {
+            String.class,
+            List.class,
+            Map.class
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/tasks/executions/Labels.java:79]: It seems overly complicated to allow multiple ways to define the label.
The original issue propose Map which is what we usually advise so I'll only propose to set labels as map

File: core/src/main/java/io/kestra/core/tasks/executions/Labels.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    private Object labels;
+
+    @Override
+    public Execution update(Execution execution, RunContext runContext) throws Exception {
+        Logger logger = runContext.logger();
+
+        Object labelsObject = labels;
+        if (labels instanceof String labelStr) {
+            try {
+                labelsObject = JacksonMapper.toObject(runContext.render(labelStr));
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/tasks/executions/Labels.java:93]: Please define the mapper as a constant as we usually do

File: core/src/test/java/io/kestra/core/tasks/flows/RuntimeLabelsTest.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            is(new Label("keyFromMap", "valueFromMap")),
+            is(new Label("keyFromList", "valueFromList")),
+            is(new Label("keyFromExecution", "valueFromExecution")),
+            is(new Label("overriddenExecutionLabelKey", labelsOverriderTaskRunId))
+        ));
+    }
+
+
+    @Test
+    void npeOnExecutionLabels() throws TimeoutException {
```
Comment (loicmathieu) [inline core/src/test/java/io/kestra/core/tasks/flows/RuntimeLabelsTest.java:55]: I didn't understand this test, it doesn't NPE?

File: core/src/main/java/io/kestra/core/tasks/executions/Labels.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+            List.class,
+            Map.class
+        }
+    )
+    @PluginProperty(dynamic = true)
+    @NotNull
+    private Object labels;
+
+    @Override
+    public Execution update(Execution execution, RunContext runContext) throws Exception {
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/tasks/executions/Labels.java:87]: The algorithm is a bit complex, if it's a string or a list you create a map, then if it's a map (which will always be the case at this point) you create a map with the existing, add the new, then create a list.

You may be able to simplify it.
And we usually check that if it's none of what we expected we throw an exception.

**Review by loicmathieu — CHANGES_REQUESTED** [review]

**Review by anna-geller — APPROVED** [review]
works great!

image


```yaml
id: labels
namespace: dev

inputs:
  - name: user
    type: STRING
    defaults: anna-geller

tasks:
  - id: update-labels-with-map
    type: io.kestra.core.tasks.executions.Labels
    labels:
      customerId: "{{inputs.user}}"

  - id: get
    type: io.kestra.core.tasks.debugs.Return
    format: https://github.com/kestra-io/kestra/issues/2632

  - id: update-list
    type: io.kestra.core.tasks.executions.Labels
    labels:
      - key: url
        value: "{{outputs.get.value}}"
```

**Comment by yuri1969** [comment]
It works nicely.

My 2 cents on the current naming:
It feels there should be emphasis on *adding* labels. Since just providing a map of `labels` in a `Labels` task makes one feel it replaces the current labels.

**Comment by anna-geller** [comment]
fair point @yuri1969 - however, this task also relaces labels when needed. Try setting default labels on the flow and then changing it from this task -- the one set on the task will prevail 🚀 

Agree we can perhaps make it even clearer in the task doc

File: core/src/main/java/io/kestra/core/tasks/executions/Labels.java (hunk 2 above)
Comment (anna-geller) [inline core/src/main/java/io/kestra/core/tasks/executions/Labels.java:38]:
```suggestion
    title = "Allow to add or overwrite labels for the current execution at runtime.",
```

File: core/src/main/java/io/kestra/core/tasks/executions/Labels.java (hunk 5 above)
Comment (anna-geller) [inline core/src/main/java/io/kestra/core/tasks/executions/Labels.java:74]:
```suggestion
        title = "Labels to add or overwrite for the current execution.",
```

File: core/src/main/java/io/kestra/core/tasks/executions/Labels.java (hunk 6 above)
Comment (brian-mulier-p) [reply core/src/main/java/io/kestra/core/tasks/executions/Labels.java:79]: I'll remove String but List is something @anna-geller want to be able to use afaik

File: core/src/main/java/io/kestra/core/tasks/executions/Labels.java (hunk 7 above)
Comment (brian-mulier-p) [reply core/src/main/java/io/kestra/core/tasks/executions/Labels.java:93]: That's already a constant behind it, I'm using a static method :+1:

File: core/src/main/java/io/kestra/core/tasks/executions/Labels.java (hunk 9 above)
Comment (brian-mulier-p) [reply core/src/main/java/io/kestra/core/tasks/executions/Labels.java:87]: I removed the String part but for the rest I'm pretty confident in the algorithm. The map part is done to easily handle overrides (as grouping them onto a single map with new values behind added afterwards will delete previous values).

**Pull request merged**